### ISSUE TITLE
Example use of collect_output()

### ISSUE THREAD

**Issue by ZodiacWind** [ronf/asyncssh#452]
Thanks for the work. I am new for asyncssh. Since I focus on time-consuming task such as machine learning, I would like to learn and use `collect_output()` to get real-time information of my remote shell.

However, I don't have much infromation about the usage of `collect_output()`, Could the author or anyone else provide som examples? I would really appreciate it.

**Comment by ronf** [comment]
The collect_output() function can be useful if you don't want to block and just wait to pick up the pending output periodically while you do something else, but you need to be careful not to call collect_output() repeatedly without doing some other blocking work in between calls, as it will just sit in a tight loop eating CPU if you do that.

If you want to run a command and read the output incrementally, you're better off using one of the flavors of read() on an SSHClientProcess than using collect_output(), as that will block if the output you're looking for is not there yet. If you want to process the output a line at a time, you can also use process.stdout as an async iterator which returns each line of output as it becomes available.

See #449 for some further discussion about this and some examples.

**Comment by ZodiacWind** [comment]
Thanks for your suggestions. My scenario is something like that I have a local service to handle outputs of multiple remote time-consuming tasks, and I want to monitor the tasks without blocking my local service. Maybe I should call `collect_output()` periodically? 
I noticed the "callbacks" in `data_received()` of `SSHClientSession`. It's nice that once new output arrives, I could handle it correspondingly, which seems suitable for my need. In addition, I also want to set `timeout` parameter for each tasks. It could be set for `run()`, but I don't find the way to set it in `SSHClientSession`. Could you give me some more suggestions please?

**Comment by ronf** [comment]
As long as your local task is written to use asyncio "await" calls when it does things that would "block", you can run that simultaneously with other tasks which are "blocking" reading out. All these tasks can run concurrently, with none of them preventing others from making process when data is available (or whatever the tasks was waiting for). See an example of this sort of parallel task setup at https://asyncssh.readthedocs.io/en/latest/#running-multiple-clients.

If you want to do everything based on callbacks rather than "blocking" read() calls, you can do that as well, but it's hard to mix the two models. So, you'd likely need to make your local service callback-based as well.

With a callback-based service, you don't really set a timeout, as you're not blocking on anything. You can however set up a time-based callback with loop.call_later() where you do something like close a session that is taking too long to produce a result. Examples of the callback-based model can be found at https://asyncssh.readthedocs.io/en/latest/#callback-example, but generally speaking this code is more complex than using the higher-level await-based API calls.

**Comment by ZodiacWind** [comment]
> Thanks for your suggestions. My scenario is something like that I have a local service to handle outputs of multiple remote time-consuming tasks, and I want to monitor the tasks without blocking my local service. Maybe I should call `collect_output()` periodically? I noticed the "callbacks" in `data_received()` of `SSHClientSession`. It's nice that once new output arrives, I could handle it correspondingly, which seems suitable for my need. In addition, I also want to set `timeout` parameter for each tasks. It could be set for `run()`, but I don't find the way to set it in `SSHClientSession`. Could you give me some more suggestions please?

Maybe I should use  something like 
```
        chan, session = await ssh_connection.create_session(MySSHClientSession, shell)
        await asyncio.wait_for(chan.wait_closed(), timeout)
```
?

**Comment by ronf** [comment]
Yes - you probably could do something like that, but in the case where the timeout occurred, the session would still be open. So, you'd want to catch the exception on that and trigger your own call to close() on the session, probably followed by another wait_closed() to wait until the close completes. Also, note that the task running the wait_for() will be blocked the whole time it is waiting for the timeout, so you'd need to create separate tasks (similar to my multiple client example above) if you wanted to do this on multiple connections at once. At that point, you'd probably better off just going back to using either run() without a timeout, or something like create_process() call if you wanted to incrementally read the output as it arrived rather than buffering it up. You could still do the wait_for() on the function doing the reading if you wanted to give up after a set time.

**Comment by ZodiacWind** [comment]
Thank you for your help. I'll try to figure it out.

**Comment by ZodiacWind** [comment]
>  So, you'd want to catch the exception on that and trigger your own call to close() on the session, probably followed by another wait_closed() to wait until the close completes

Could I dispose the SSH session by force, if `wait_closed()` got stuck?

**Comment by ronf** [comment]
If you call close() from another task, the task calling wait_closed() will be unblocked, looking like it returned from the wait_closed(). If you wrap a wait_for() around the wait_closed(), it'll raise the asyncio timeout exception if the connection hasn't already closed after that amount of time, at which point you can then call close() from the same task. After that, you can call wait_closed() again to wait for the connection to clean itself up, but you can expect at that point that the new wait_closed() should not take long.

**Comment by ZodiacWind** [comment]
Thank you. I tried the following codes, which seem to work. Though I am not sure whether I have caught your point.

```
    async with asyncssh.connect(
            host=host,
            port=port,
            username=username,
            client_keys=[private_key],
            known_hosts=None) as ssh_connection:

        chan, session = await ssh_connection.create_session(MySSHClientSession, shell)
        try:
            await asyncio.wait_for(chan.wait_closed(), timeout)
        except asyncio.TimeoutError:
            session.success = False
            session.err += 'Timeout error: The task takes too long...'
            ssh_connection.close()
            await chan.wait_closed()
```

Is it necessary to call `await chan.wait_closed()` after  `ssh_connection.close()`?

**Comment by ronf** [comment]
Since you are closing the connection here, you'd normally want the last wait_closed() call to also be on the connection rather than the channel. However, since you are using "async with" on the call to connect(), you don't actually need either the ssh_connection.close() or a call to ssh_connection.wait_closed(). When you exit out of the "async with", AsyncSSH will automatically do both of those things for you.

You don't need a wait_closed() on the channel at all in this case, as waiting for the connection to close will also implicitly wait for all the channels to close, so the "async with" takes care of that for you as well.

**Comment by ZodiacWind** [comment]
Got it. Thank you for your help. Really RESPONSIVE!

**Comment by ronf** [comment]
My pleasure - happy to help!
